eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -200,10 +204,43 @@ class NodeRelay private(nodeParams: NodeParams,
             rejectPayment(upstream, Some(failure))
             stopping()
           case None =>
-            doSend(upstream, nextPayload, nextPacket)
+            if (nextPayload.isAsyncPayment && nodeParams.features.hasFeature(Features.AsyncPaymentPrototype)) {
+              waitForTrigger(upstream, nextPayload, nextPacket)
+            } else {
+              doSend(upstream, nextPayload, nextPacket)
+            }
         }
     }
 
+  private def waitForTrigger(upstream: Upstream.Trampoline, nextPayload: IntermediatePayload.NodeRelay.Standard, nextPacket: OnionRoutingPacket): Behavior[Command] = {
+    context.log.info(s"waiting for async payment to trigger before relaying trampoline payment (amountIn=${upstream.amountIn} expiryIn=${upstream.expiryIn} amountOut=${nextPayload.amountToForward} expiryOut=${nextPayload.outgoingCltv}, asyncPaymentsParams=${nodeParams.relayParams.asyncPaymentsParams})")
+    // a trigger must be received before waiting more than `holdTimeoutBlocks`
+    val timeoutBlock: BlockHeight = nodeParams.currentBlockHeight + nodeParams.relayParams.asyncPaymentsParams.holdTimeoutBlocks
+    // a trigger must be received `cancelSafetyBeforeTimeoutBlocks` before the incoming payment cltv expiry
+    val safetyBlock: BlockHeight = (upstream.expiryIn - nodeParams.relayParams.asyncPaymentsParams.cancelSafetyBeforeTimeout).blockHeight
+    val messageAdapter = context.messageAdapter[CurrentBlockHeight](cbc => WrappedCurrentBlockHeight(cbc.blockHeight))
+    context.system.eventStream ! EventStream.Subscribe[CurrentBlockHeight](messageAdapter)
+
+    // TODO: send the WaitingToRelayPayment message to an actor that watches for the payment receiver to come back online before sending the RelayAsyncPayment message
+    context.system.eventStream ! EventStream.Publish(WaitingToRelayPayment(nextPayload.outgoingNodeId, paymentHash))
+    Behaviors.receiveMessagePartial {
+      case WrappedCurrentBlockHeight(blockHeight) if blockHeight >= safetyBlock =>
+        context.log.warn(s"rejecting async payment at block $blockHeight; was not triggered ${nodeParams.relayParams.asyncPaymentsParams.cancelSafetyBeforeTimeout} safety blocks before upstream cltv expiry at ${upstream.expiryIn}")
+        rejectPayment(upstream, Some(TemporaryNodeFailure)) // TODO: replace failure type when async payment spec is finalized
+        stopping()
+      case WrappedCurrentBlockHeight(blockHeight) if blockHeight >= timeoutBlock =>
+        context.log.warn(s"rejecting async payment at block $blockHeight; was not triggered after waiting ${nodeParams.relayParams.asyncPaymentsParams.holdTimeoutBlocks} blocks")
+        rejectPayment(upstream, Some(TemporaryNodeFailure)) // TODO: replace failure type when async payment spec is finalized
+        stopping()
```

Comment:
We should handle all `WrappedCurrentBlockHeight` cases, otherwise we'll have a warning in the logs at every block saying we didn't handle `WrappedCurrentBlockHeight`, which will be annoying. Just add the following:

```scala
      case WrappedCurrentBlockHeight(blockHeight) => 
        Behaviors.same
```

Reply:
Fixed in [4c0b4d2](https://github.com/ACINQ/eclair/pull/2435/commits/4c0b4d281e1a5386cb516cd2cf59b9a56bbdab99).

Am I correct in assuming that once we pass out of the `waitForTrigger` state, that the next state won't get any more `WrappedCurrentBlockHeight` messages because `messageAdapter` that they were being sent to will have gone out of scope?

Reply:
That's a good comment, I'm not sure...this is worth testing and potentially fixing before merging this PR, good catch! Otherwise I think there is a way to tell the event stream that we "unregister".

Reply:
There is an `EventStream.Unsubscribe` message available, it's probably a good idea to send that when moving out of the waiting state (no need to test it though).